### ISSUE TITLE
ST10: False positive on Redundant constant expression

### ISSUE THREAD

**Issue by FrankTub** [sqlfluff/sqlfluff#6492]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I upgraded to SQLFluff `3.3.0` and now below part of one of my queries reports an error where there. is none:

```sql
        case
            when (rank() over (
                partition by category_id
                order by year_count, month_count
            ) - 1) % 12 = 0
```

The error is on the `=` after 12 and it returns:

```shell
L: 153 | P: 43 | ST10 | Redundant constant expression. [structure.constant_expression]
```

Given that the `rank() over (..)` does not return a static value the evaluation of that line is also not static.

### Expected Behaviour

Not have an error in this scenario

### Observed Behaviour

See What Happened.

### How to reproduce

I tried to make a minimal case:

```sql
select case when 1 % 12 = 0 then 1 end as test
```

Not really sure if the ST10 is supposed to pick up this statement or not given that now my input is indeed a constant and could be simplified. So what I think is maybe a better minimal test case is:

```sql
with
cte as (
    select 1 as num
    union all
    select 2 as num
    union all
    select 3 as num
    union all
    select 4 as num
)

select
    *,
    case
        when num % 2 = 0 then 1
        else 2
    end as test
from cte
```

### Dialect

dbt-postgres

### Version

```shell
sqlfluff --version           
sqlfluff, version 3.3.0
```

```shell
python --version               
Python 3.11.9
```

### Configuration

My `.sqlfluff`:

```
[sqlfluff]
# Supported dialects https://docs.sqlfluff.com/en/stable/dialects.html
# Or run 'sqlfluff dialects'
dialect = postgres
# One of [raw|jinja|python|placeholder]
templater = dbt
# Comma separated list of rules to exclude, or None
exclude_rules = L031
# Below controls SQLFluff output, see max_line_length for SQL output
output_line_length = 160
max_line_length = 160
# Very large files can make the parser effectively hang.
# The more efficient check is the _byte_ limit check which
# is enabled by default. The previous _character_ limit check
# is still present for backward compatibility. This will be
# removed in a future version.
# Set either to 0 to disable.
large_file_skip_byte_limit = 50000

[sqlfluff:templater:dbt]
project_dir = .
profiles_dir = ~/.dbt

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:rules]
single_table_references = unqualified

[sqlfluff:rules:capitalisation.keywords]
# Keywords
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.identifiers]
# Unquoted identifiers
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.functions]
# Function names
extended_capitalisation_policy = lower

[sqlfluff:rules:capitalisation.literals]
# Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:capitalisation.types]
# Data Types
extended_capitalisation_policy = lower

[sqlfluff:rules:layout.long_lines]
# Line length
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:rules:references.keywords]
# Comma separated list of words to ignore for this rule
ignore_words = date,year_month,month,name,attribute,version,level,path,bit

[sqlfluff:rules:structure.subquery]
# By default, allow subqueries in from clauses, but not join clauses
forbid_subquery_in = both

[sqlfluff:rules:ambiguous.join]
fully_qualify_join_types = both

[sqlfluff:rules:ambiguous.column_references]
# GROUP BY/ORDER BY column references
group_by_and_order_by_style = explicit

[sqlfluff:rules:references.special_chars]
# Special characters in identifiers
additional_allowed_characters = ""

[sqlfluff:rules:convention.casting_style]
# SQL type casting
preferred_type_casting_style = cast

[sqlfluff:layout:type:select_clause]
line_position = strict

[sqlfluff:layout:type:where_clause]
line_position = strict

[sqlfluff:layout:type:from_clause]
line_position = strict

[sqlfluff:layout:type:join_clause]
line_position = strict

[sqlfluff:layout:type:groupby_clause]
line_position = strict

[sqlfluff:layout:type:having_clause]
line_position = strict

```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by rogalski** [comment]
Thanks for bug report.

This is really odd, let me reproduce locally and get back to you.
